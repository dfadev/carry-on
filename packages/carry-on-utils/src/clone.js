/** @format **/
export default function clone(value) {
  if (Array.isArray(value)) return value.slice();

  if (value instanceof Object) return { ...value };

  return value;
}
