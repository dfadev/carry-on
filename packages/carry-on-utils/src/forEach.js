import keys from "./keys";

export function forOwn(object, iteratee) {
  for (let i = 0, ks = keys(object), len = ks.length; i < len; i += 1) {
    const key = ks[i];
    const val = object[key];
    const result = iteratee(val, key);
    if (result === false) break;
  }

  return object;
}

export default function forEach(collection, iteratee) {
  if (Array.isArray(collection)) {
    for (let i = 0, len = collection.length; i < len; i += 1)
      if (iteratee(collection[i], i) === false) break;
  } else {
    forOwn(collection, iteratee);
  }
  return collection;
}
