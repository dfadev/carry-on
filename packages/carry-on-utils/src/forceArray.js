/** @format **/
export default function forceArray(thing) {
  return Array.isArray(thing) ? thing : [thing];
}
