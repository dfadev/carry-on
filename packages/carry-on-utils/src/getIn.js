import toPath from "./toPath";

export function getInA(state, path, def) {
  let current = state;
  for (let i = 0; i < path.length; i += 1) {
    const key = path[i];
    if (
      current === undefined ||
      current === null ||
      typeof current !== "object" ||
      (Array.isArray(current) && Number.isNaN(key))
    ) {
      return def;
    }
    current = current[key];
  }
  return current === undefined ? def : current;
}

export default function getIn(state, complexKey, def) {
  return getInA(state, toPath(complexKey), def);
}
