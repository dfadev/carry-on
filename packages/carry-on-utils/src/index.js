/** @format **/
export { throttle, debounce } from "throttle-debounce";
export { default as debouncePromise } from "debounce-promise";
export { default as isEqual } from "fast-deep-equal";
export { default as mutateMerge } from "./mutateMerge";

export { default as isFunction } from "./isFunction";
export { default as toPath } from "./toPath";
export { default as getIn, getInA } from "./getIn";
export { default as makeCancelable } from "./makeCancelable";
export { default as setIn } from "./setIn";
export { default as mutateSet, mutateSetA } from "./mutateSet";
export * from "./proxyState";
