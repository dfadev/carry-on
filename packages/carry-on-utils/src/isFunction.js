/** @format **/
// is the thing a function
export default function isFunction(thing) {
  return !!(thing && thing.constructor && thing.call && thing.apply);
}
