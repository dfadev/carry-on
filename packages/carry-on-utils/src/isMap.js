export default function isMap(target) {
	return target instanceof Map
}
