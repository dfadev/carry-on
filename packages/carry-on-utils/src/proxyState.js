/** @format **/
// original implementation from proxyequal
import { getCollectionHandlers, shouldInstrument } from "./shouldInstrument";
import mutateSet from "./mutateSet";

const hasProxy = typeof Proxy !== "undefined";
/* istanbul ignore next */
if (!hasProxy) throw new Error("Missing Proxy support");
const ProxyConstructor = Proxy;

const ProxyToState = new WeakMap();

export const isProxyfied = object =>
  object && typeof object === "object" ? ProxyToState.has(object) : false;

export const deproxify = object =>
  (object !== undefined && typeof object === "object"
    ? ProxyToState.get(object)
    : object) || object;

const prepareObject = state => {
  if (!Object.isFrozen(state)) return state;

  // unfreeze
  if (Array.isArray(state)) return state.slice(0);
  if (state.constructor.name === "Object") {
    const clone = { ...state };
    Object.setPrototypeOf(clone, Object.getPrototypeOf(state));
    return clone;
  }
  return state;
};

const shouldProxy = type => type === "object";

function proxyfy(state, onGet, suffix, ProxyMap) {
  if (!state) return state;

  const alreadyProxy = isProxyfied(state);

  if (!alreadyProxy && !shouldInstrument(state)) return state;

  const hasCollectionHandlers = !alreadyProxy && getCollectionHandlers(state);

  const storedValue = ProxyMap.get(state) || {};

  if (storedValue[suffix]) return storedValue[suffix];

  const theBaseObject = alreadyProxy ? state : prepareObject(state);

  let proxyValue;

  function iterable(key, iterator) {
    let index = 0;
    const next = () => {
      const nextItem = iterator.next();
      const subKey = key + "." + index;
      index++;
      return {
        ...nextItem,
        get value() {
          if (nextItem.done && !nextItem.value) {
            return undefined;
          }
          return proxyValue(subKey, nextItem.value);
        }
      };
    };
    return {
      [Symbol.iterator]: () => ({
        next
      }),
      next
    };
  }

  proxyValue = (key, value) => {
    // suffix should be prefix???
    const thisId = suffix ? suffix + '["' + key + '"]' : key;
    const type = typeof value;

    onGet(thisId);

    if (shouldProxy(type)) {
      return proxyfy(value, onGet, thisId, ProxyMap);
    }

    if (hasCollectionHandlers) {
      switch (key) {
        case "get":
          return k => proxyValue(k, state.get(k));
        case "has":
          return k => proxyValue(k, state.has(k));
        case "keys":
          return () => state.keys();
        case "values":
          return () => iterable(key, state.values());
        case "entries":
          return () => iterable(key, state.entries());
        /* istanbul ignore next */
        case [Symbol.iterator]:
          // this is broken
          return iterable(key, state[Symbol.iterator].bind(state));
        default:
      }
    }

    return value;
  };

  const hooks = {
    set() {
      throw new Error("can't mutate state here");
    },

    get(target, prop) {
      const storeValue = state[prop];
      return typeof prop === "string"
        ? proxyValue(prop, storeValue)
        : storeValue;
    }
  };

  const proxy = new ProxyConstructor(theBaseObject, hooks);
  storedValue[suffix] = proxy;
  ProxyMap.set(state, storedValue);
  ProxyToState.set(proxy, state);
  return proxy;
}

export const proxyState = (object, _ProxyMap) => {
  const affected = {};
  const keysUsed = new Set();
  const ProxyMap = _ProxyMap || new WeakMap();
  let sealed = false;

  const onKeyUse = key => {
    if (sealed) return;
    if (!keysUsed.has(key)) {
      keysUsed.add(key);
      mutateSet(affected, key, true);
    }
  };

  const createState = state => proxyfy(state, onKeyUse, "", ProxyMap);

  return {
    affected,
    state: createState(object),
    seal: () => (sealed = true),
    unseal: () => (sealed = false),
    reset: () => {
      affected.length = 0;
      keysUsed.clear();
    },
    replaceState(state) {
      this.state = createState(state);
      this.unseal();
      return this;
    }
  };
};
