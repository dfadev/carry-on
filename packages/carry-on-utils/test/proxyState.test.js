import { proxyState, deproxify, isProxyfied } from "../src/proxyState";

describe("proxy", () => {
  it("arrays", () => {
    const A1 = [0, 1, 2, 3];
    const A2 = [0, 1, 2, 3];
    const A3 = A1.map(a => a);
    const A4 = A1.map(a => a * 2);

    const trapped = proxyState(A1);

    expect(trapped.affected).toMatchSnapshot();

    trapped.state[0]; // += 0;
    trapped.state[0]; // += 0;

    expect(trapped.affected).toMatchSnapshot();

    trapped.state[1]; // += 0;

    expect(trapped.affected).toMatchSnapshot();

    trapped.seal();
    trapped.state[2]; // += 0;
    expect(trapped.affected).toMatchSnapshot();
    trapped.unseal();

    trapped.state[3]; // += 0;
    expect(trapped.affected).toMatchSnapshot();
  });

  it("objects", () => {
    const A1 = {
      key1: 1,
      key2: {
        array: [1, 2, 4]
      },
      key3: null
    };
    const A2 = {
      key1: 1,
      key2: 2
    };
    const A3 = { ...A1, key1: 2 };
    const A4 = {
      key1: 1,
      key2: {
        array: [1, 2, 4]
      }
    };

    const trapped = proxyState(A1);
    expect(trapped.affected).toMatchSnapshot();

    trapped.state.key1; // += 0;
    trapped.state.key1; // += 0;
    expect(trapped.affected).toMatchSnapshot();

    trapped.reset();
    expect(trapped.affected).toMatchSnapshot();
    trapped.state.key2.array[0]; // += 0;
    expect(trapped.affected).toMatchSnapshot();
  });

  it("handle empty case", () => {
    const A = {
      a: undefined
    };
    expect(proxyState(A).state.a).toMatchSnapshot();
    expect(proxyState(undefined).state).toMatchSnapshot();
  });

  it("plain shallowEqual", () => {
    const A1 = {
      key1: 1,
      key2: {
        array: [1, 2, 4]
      },
      key3: null
    };
    const A2 = {
      key1: 1,
      key2: 2
    };
    const A3 = { ...A1, key1: 2 };
    const A4 = {
      key1: 1,
      key2: A1.key2
    };

    const trapped = proxyState(A1);
    expect(trapped.affected).toMatchSnapshot();

    trapped.state.key1; // += 0;
    trapped.state.key1; /// += 0;
    expect(trapped.affected).toMatchSnapshot();

    trapped.reset();
    expect(trapped.affected).toMatchSnapshot();
    trapped.state.key2.array[0]; // += 0;
    expect(trapped.affected).toMatchSnapshot();
  });

  it("nested shallowEqual", () => {
    const A1 = {
      key1: 1,
      key2: {
        array: [1, 2, 4]
      },
      key3: null
    };
    const A2 = {
      key1: 1,
      key2: {
        array: A1.key2.array
      }
    };

    const trapped1 = proxyState(A1);
    const trapped2 = proxyState(trapped1.state);

    trapped2.state.key1; // += 0;
    expect(trapped1.affected).toMatchSnapshot();
    expect(trapped2.affected).toMatchSnapshot();

    trapped2.state.key1; // += 1;
    A1.key1 += 1;

    trapped1.reset();
    trapped2.reset();
    trapped2.state.key2.array[0]; // += 0;
    expect(trapped1.affected).toMatchSnapshot();
  });

  it("can proxy via proxy", () => {
    const A = {
      b: {
        c: {
          d: 1
        }
      }
    };
    const p1 = proxyState(A);
    const p2 = proxyState(p1.state.b);

    expect(p1.state).toMatchSnapshot();
    expect(p2.state).toMatchSnapshot();
  });

  it("track proxy via proxy", () => {
    const A = {
      root: {
        b: {
          a: 1,
          b: true,
          c: {
            d: 1,
            e: 14
          }
        }
      }
    };
    const p0 = proxyState(A);
    const p1 = proxyState(p0.state.root);
    const p2 = proxyState(p1.state.b);
    p2.state.c.d; // ++;

    expect(p1.affected).toMatchSnapshot();
    expect(p2.affected).toMatchSnapshot();
    p2.state.a; // ++;

    expect(p1.affected).toMatchSnapshot();
    expect(p2.affected).toMatchSnapshot();

    expect(p1.affected).toMatchSnapshot();
    expect(p2.affected).toMatchSnapshot();

    expect(p0.affected).toMatchSnapshot();
  });

  it("shallow equal test", () => {
    const C = { c: 1 };
    const A1 = {
      b: C
    };
    const A2 = {
      b: C
    };

    expect(proxyState(A1).state.b).not.toBe(proxyState(A2).state.b);

    const p1 = proxyState(A1);
    const s1 = p1.state;
    const p2 = p1.replaceState(A2);
    const s2 = p2.state;

    expect(s1.b).toBe(s2.b);
  });

  it("handles freezed objects", () => {
    const O1 = {
      a: 1,
      b: 2,
      c: { d: 4 }
    };
    const O2 = Object.freeze(O1);

    const trapped = proxyState(O2);
    const { state } = trapped;
    const read = state.a + state.b + state.c.d;
    expect(read).toEqual(7);
    expect(trapped.affected).toMatchSnapshot();
  });

  it("handles prototype chain", () => {
    const o1 = {
      a1: 1
    };
    const o2 = {
      a2: 2
    };
    Object.setPrototypeOf(o2, o1);
    expect(Object.getPrototypeOf(o2)).toBe(o1);
    expect(o1.isPrototypeOf(o2)).toBe(true);
    const trapped = proxyState(o2);
    expect(o1.isPrototypeOf(trapped.state)).toBe(true);
  });

  it("detect self", () => {
    const A = { a: 1 };
    const B = proxyState(A).state;
    const C = deproxify(B);

    expect(isProxyfied(A)).toBe(false);
    expect(isProxyfied(B)).toBe(true);
    expect(isProxyfied(C)).toBe(false);
    expect(C).toBe(A);
  });

  describe("types", () => {
    it("should handle date", () => {
      const A = { d: new Date() };
      const p = proxyState(A);
      const B = p.state;
      expect(B.d.getDate()).toEqual(new Date().getDate());
      expect(p.affected).toMatchSnapshot();
    });

    it("should handle Promise", () => {
      const A = { d: Promise.resolve() };
      const p = proxyState(A);
      const B = p.state;
      const result = B.d.then(() => true);
      expect(p.affected).toMatchSnapshot();
      return result;
    });

    it("should handle Map.get", () => {
      const A = { d: new Map([[1, 2]]) };
      const p = proxyState(A);
      const B = p.state;
      expect(B.d.get(1)).toBe(2);
      expect(p.affected).toMatchSnapshot();
    });

    it("should handle Map.keys", () => {
      const A = { d: new Map([["key", "value"]]) };
      const p = proxyState(A);
      const B = p.state;
      expect([...B.d.keys()]).toEqual([...A.d.keys()]);
      expect(p.affected).toMatchSnapshot();
    });

    it("should handle Map.values", () => {
      const A = { d: new Map([[1, 2]]) };
      const p = proxyState(A);
      const B = p.state;
      expect([...B.d.values()]).toEqual([...A.d.values()]);
      expect(p.affected).toMatchSnapshot();
    });

    it("should handle Map.entries", () => {
      const A = { d: new Map([[1, { sub1: 1, sub2: 2 }]]) };
      const p = proxyState(A);
      const B = p.state;

      expect(B.d.entries().next().value[1].sub2).toBe(2);
      expect(p.affected).toMatchSnapshot();

      expect([...B.d.entries()]).toEqual([...A.d.entries()]);
      expect(p.affected).toMatchSnapshot();
    });

    it("should handle Map.has", () => {
      const A = { d: new Map([[1, 2]]) };
      const p = proxyState(A);
      const B = p.state;
      expect(B.d.has(1)).toBe(true);
      expect(p.affected).toMatchSnapshot();
    });

    it("should handle Map.iterator", () => {
      const A = { d: new Map([[1, 2]]) };
      const p = proxyState(A);
      const B = p.state;

      const realMap = [];
      for (const item of new Map([[1, 2]])) realMap.push(item);

      const proxyMap = [];
      //console.log(B.d.entries(), B.d.values());
      //console.log(typeof new Map(), typeof B.d, B.d, B.d.entries);
      for (const item of B.d) proxyMap.push(item);

      expect(proxyMap).toEqual(realMap);
    });
  });

  describe("set", () => {
    it("should not let you set value", () => {
      const A = { a: 1, b: 2 };
      const p = proxyState(A);
      const B = p.state;

      expect(() => (B.a = 2)).toThrow();
      expect(A.a).toBe(1);
      expect(B.a).toBe(1);
    });
  });

  it("isProxified(number/string) should return false", () => {
    expect(isProxyfied(1)).toBe(false);
    expect(isProxyfied("1")).toBe(false);
  });

  it("deproxify", () => {
    expect(deproxify({ a: 1 })).toMatchSnapshot();
    const p = proxyState({ a: new Map([[1, { b: 2 }]]) });
    expect(deproxify(p.state.a)).toMatchSnapshot();
  });

  describe("corners", () => {
    it("undefined", () => {
      const p = proxyState(undefined);
      expect(p).toMatchSnapshot();
    });

    it("null", () => {
      const p = proxyState(null);
      expect(p).toMatchSnapshot();
    });

    it("number", () => {
      const p = proxyState(99);
      expect(p).toMatchSnapshot();
    });

    it("string", () => {
      const p = proxyState("abc");
      expect(p).toMatchSnapshot();
    });

    it("array", () => {
      const p = proxyState([1,2,3]);
      expect(p).toMatchSnapshot();
    });

    it("deproxify(undefined)", () => {
      const p = deproxify(undefined);
      expect(p).toMatchSnapshot();
    });

    it("deproxify(null)", () => {
      const p = deproxify(null);
      expect(p).toMatchSnapshot();
    });
  });
});
