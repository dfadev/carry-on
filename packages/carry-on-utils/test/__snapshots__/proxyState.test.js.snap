// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`proxy arrays 1`] = `Object {}`;

exports[`proxy arrays 2`] = `
Object {
  "0": true,
}
`;

exports[`proxy arrays 3`] = `
Object {
  "0": true,
  "1": true,
}
`;

exports[`proxy arrays 4`] = `
Object {
  "0": true,
  "1": true,
}
`;

exports[`proxy arrays 5`] = `
Object {
  "0": true,
  "1": true,
  "3": true,
}
`;

exports[`proxy can proxy via proxy 1`] = `
Object {
  "b": Object {
    "c": Object {
      "d": 1,
    },
  },
}
`;

exports[`proxy can proxy via proxy 2`] = `
Object {
  "c": Object {
    "d": 1,
  },
}
`;

exports[`proxy corners array 1`] = `
Object {
  "affected": Object {},
  "replaceState": [Function],
  "reset": [Function],
  "seal": [Function],
  "state": Array [
    1,
    2,
    3,
  ],
  "unseal": [Function],
}
`;

exports[`proxy corners deproxify(null) 1`] = `null`;

exports[`proxy corners deproxify(undefined) 1`] = `undefined`;

exports[`proxy corners null 1`] = `
Object {
  "affected": Object {},
  "replaceState": [Function],
  "reset": [Function],
  "seal": [Function],
  "state": null,
  "unseal": [Function],
}
`;

exports[`proxy corners number 1`] = `
Object {
  "affected": Object {},
  "replaceState": [Function],
  "reset": [Function],
  "seal": [Function],
  "state": 99,
  "unseal": [Function],
}
`;

exports[`proxy corners string 1`] = `
Object {
  "affected": Object {},
  "replaceState": [Function],
  "reset": [Function],
  "seal": [Function],
  "state": "abc",
  "unseal": [Function],
}
`;

exports[`proxy corners undefined 1`] = `
Object {
  "affected": Object {},
  "replaceState": [Function],
  "reset": [Function],
  "seal": [Function],
  "state": undefined,
  "unseal": [Function],
}
`;

exports[`proxy deproxify 1`] = `
Object {
  "a": 1,
}
`;

exports[`proxy deproxify 2`] = `
Map {
  1 => Object {
    "b": 2,
  },
}
`;

exports[`proxy handle empty case 1`] = `undefined`;

exports[`proxy handle empty case 2`] = `undefined`;

exports[`proxy handles freezed objects 1`] = `
Object {
  "a": true,
  "b": true,
  "c": Object {
    "d": true,
  },
}
`;

exports[`proxy nested shallowEqual 1`] = `
Object {
  "key1": true,
}
`;

exports[`proxy nested shallowEqual 2`] = `
Object {
  "key1": true,
}
`;

exports[`proxy nested shallowEqual 3`] = `
Object {
  "key1": true,
  "key2": Object {
    "array": Object {
      "0": true,
    },
  },
  "length": 0,
}
`;

exports[`proxy objects 1`] = `Object {}`;

exports[`proxy objects 2`] = `
Object {
  "key1": true,
}
`;

exports[`proxy objects 3`] = `
Object {
  "key1": true,
  "length": 0,
}
`;

exports[`proxy objects 4`] = `
Object {
  "key1": true,
  "key2": Object {
    "array": Object {
      "0": true,
    },
  },
  "length": 0,
}
`;

exports[`proxy plain shallowEqual 1`] = `Object {}`;

exports[`proxy plain shallowEqual 2`] = `
Object {
  "key1": true,
}
`;

exports[`proxy plain shallowEqual 3`] = `
Object {
  "key1": true,
  "length": 0,
}
`;

exports[`proxy plain shallowEqual 4`] = `
Object {
  "key1": true,
  "key2": Object {
    "array": Object {
      "0": true,
    },
  },
  "length": 0,
}
`;

exports[`proxy track proxy via proxy 1`] = `
Object {
  "b": Object {
    "c": Object {
      "d": true,
    },
  },
}
`;

exports[`proxy track proxy via proxy 2`] = `
Object {
  "c": Object {
    "d": true,
  },
}
`;

exports[`proxy track proxy via proxy 3`] = `
Object {
  "b": Object {
    "a": true,
    "c": Object {
      "d": true,
    },
  },
}
`;

exports[`proxy track proxy via proxy 4`] = `
Object {
  "a": true,
  "c": Object {
    "d": true,
  },
}
`;

exports[`proxy track proxy via proxy 5`] = `
Object {
  "b": Object {
    "a": true,
    "c": Object {
      "d": true,
    },
  },
}
`;

exports[`proxy track proxy via proxy 6`] = `
Object {
  "a": true,
  "c": Object {
    "d": true,
  },
}
`;

exports[`proxy track proxy via proxy 7`] = `
Object {
  "root": Object {
    "b": Object {
      "a": true,
      "c": Object {
        "d": true,
      },
    },
  },
}
`;

exports[`proxy types should handle Map.entries 1`] = `
Object {
  "d": Object {
    "entries": true,
    "entries.0": Object {
      "1": Object {
        "sub2": true,
      },
    },
  },
}
`;

exports[`proxy types should handle Map.entries 2`] = `
Object {
  "d": Object {
    "entries": true,
    "entries.0": Object {
      "0": true,
      "1": Object {
        "asymmetricMatch": true,
        "nodeType": true,
        "sub1": true,
        "sub2": true,
      },
      "asymmetricMatch": true,
      "length": true,
      "nodeType": true,
    },
  },
}
`;

exports[`proxy types should handle Map.get 1`] = `
Object {
  "d": Object {
    "1": true,
    "get": true,
  },
}
`;

exports[`proxy types should handle Map.has 1`] = `
Object {
  "d": Object {
    "1": true,
    "has": true,
  },
}
`;

exports[`proxy types should handle Map.keys 1`] = `
Object {
  "d": Object {
    "keys": true,
  },
}
`;

exports[`proxy types should handle Map.values 1`] = `
Object {
  "d": Object {
    "values": true,
    "values.0": true,
  },
}
`;

exports[`proxy types should handle Promise 1`] = `
Object {
  "d": true,
}
`;

exports[`proxy types should handle date 1`] = `
Object {
  "d": true,
}
`;
